http: add "Connection: close" header to final server response.

m-szymon opened the pull request:
According to HTTP/1.1 standard (RFC9112 chapter 9.6 Tear-down):
> **A sender SHOULD send a Connection header field containing the "close" connection option when it intends to close a connection.**
> [...] in a response, the same field indicates that the server is going to close this connection after the response message is complete. [...]
> A server that receives a "close" connection option MUST initiate closure of the connection (see below) after it sends the final response to the request that contained the "close" connection option. **The server SHOULD send a "close" connection option in its final response on that connection.**

Fixes #3041
Refs scylladb/scylladb#26298

nyh commented:
Thanks for finding where the RFC dictates this behavior (even if just a SHOULD and not a MUST). For people who are curious exactly which RFC this is (as I was), it is https://www.rfc-editor.org/rfc/rfc9112.html#name-tear-down

Indeed,

> A server that receives a "close" connection option MUST initiate closure of the connection (see below) after it sends the final response to the request that contained the "close" connection option. **The server SHOULD send a "close" connection option in its final response on that connection.** The server MUST NOT process any further requests received on that connection.

nyh commented on src/http/httpd.cc:
```diff
@@ -183,11 +183,24 @@ set_request_content(std::unique_ptr<http::request> req, input_stream<char>* cont
+static void set_keep_alive_header(http::reply& resp, bool keep_alive) {
+    if (keep_alive) {
+        if (resp._version == "1.0") {
+            resp.add_header("Connection", "Keep-Alive");
```
Looks good, but worthy of a comment: This case is NOT standard HTTP 1.0, but rather a non-standard extension over HTTP 1.0 which is documented here https://datatracker.ietf.org/doc/html/rfc2068#section-19.7.1. 

Anyway, since this is correct and also appears to be pre-existing code that you moved, I can't complain.

m-szymon commented:
I added unit test and changed the function name as requested.

> > If the 2nd issue is in Scylla, then corresponding seastar update SHOULD(*) include the corresponding Fixes tag
> 
> Exactly. The Seastar patch should "Fixes" the Seastar issue. Later, in the ScyllaDB repository when you update the Seastar submodule you should "Fixes" the ScyllaDB bug.

But that doesn't exclude referring to ScyllaDb issue here, which was the main question. I guess comment in pull request is not too much: https://github.com/scylladb/scylladb/issues/26298

avikivity commented:
> I added unit test and changed the function name as requested.
> 
> > > If the 2nd issue is in Scylla, then corresponding seastar update SHOULD(*) include the corresponding Fixes tag
> > 
> > 
> > Exactly. The Seastar patch should "Fixes" the Seastar issue. Later, in the ScyllaDB repository when you update the Seastar submodule you should "Fixes" the ScyllaDB bug.
> 
> But that doesn't exclude referring to ScyllaDb issue here, which was the main question. I guess comment in pull request is not too much: [scylladb/scylladb#26298](https://github.com/scylladb/scylladb/issues/26298)

As I said, it's good practice to refer to the ScyllaDB issue, even in the patch changelog. Just don't mark it as "Fixes".

nyh commented:
Thanks, I'll merge this patch, and then submit a PR for ScyllaDB to update Seastar.
